app/services/merge_collection.rb

```diff
@@ -0,0 +1,221 @@
+# frozen_string_literal: true
+
+class MergeCollection
+  class << self
+    def call(uuid, delete_collection: true)
+      instance = new(uuid, delete_collection)
+      instance.save
+      instance
+    end
+  end
+
+  attr_reader :uuid,
+              :delete_collection,
+              :errors
+
+  def initialize(uuid, delete_collection)
+    @uuid = uuid
+    @delete_collection = delete_collection
+    @errors = []
+    @successful = false # updated by #save
+  end
+
+  def save
+    return false if collection.nil?
+    return false unless valid?
+
+    ActiveRecord::Base.transaction do
+      PaperTrail.request(whodunnit: paper_trail_user) do
+        new_work = merge_collection
+
+        if !new_work.valid?
+          new_work.errors.full_messages.each { |err| errors << err }
+          return false
+        end
+
+        new_work.save!
+
+        new_work.update_index
+      end
+
+      if delete_collection?
+        collection.destroy!
+
+        collection.works.each do |work|
+          work.versions.each do |version|
+            DestroyWorkVersion.call(version, force: true)
+          end
+        end
+      end
+    end
+
+    @successful = true
+    true
+  end
+
+  def delete_collection?
+    @delete_collection
+  end
+
+  def valid?
+    validate
+    errors.empty?
+  end
+
+  def successful?
+    @successful
+  end
+
+  def validate
+    collection_works = collection.works
+    canonical_metadata = work_metadata(canonical_work)
+
+    collection_works.each do |w|
+      w_metadata = work_metadata(w)
+      work_metadata_diff = MetadataDiff.call(
+        OpenStruct.new(metadata: canonical_metadata),
+        OpenStruct.new(metadata: w_metadata)
+      )
+
+      if w_metadata[:num_work_versions] != 1
+        errors << "Work-#{w.id} has #{w_metadata[:num_work_versions]} work versions, but must only have 1"
+      end
+
+      if w_metadata[:num_files] != 1
+        errors << "Work-#{w.id} has #{w_metadata[:num_files]} files, but must only have 1"
+      end
+
+      if work_metadata_diff.any?
+        diff = work_metadata_diff.inspect
+        errors << "Work-#{canonical_work.id} has different work metadata than Work-#{w.id}\n#{diff}"
+      end
+
+      if !w.representative_version.published?
+        errors << "Work-#{canonical_work.id} is not published"
+      end
```

Comment:
@nicktiberi wondering if it makes sense to move these to `validate_work(w)` method and also move line 70, 71 in that method too?

Reply:
I like that idea - this might even make Code Climate a bit happier 🙂